async_simple/coro/FutureAwaiter.h

```diff
@@ -31,29 +33,81 @@ namespace coro::detail {
 template <typename T>
 struct FutureAwaiter {
     Future<T> future_;
+    std::shared_ptr<Future<T>> sharedFuture_;
```

Comment:
Let's use union or variant for future_ and sharedFuture_.  Please add a comment for it.